### PR TITLE
Runtime sysinfo check

### LINKED ISSUE #15159: container hostConfig not checked on container start
On container start, exactly in `daemon/start.go`, only the `hostConfig` provided to the API is checked.
If you've previously created a container with a specific hostConfig and then you make changes to the system something could be broken.
I've hit this when unmounting cgroups (memory in my case) but any other hostConfig's field could have this problem I think.
Step to reproduce on an Ubuntu 15.04:
```bash
$ docker run -d --memory-swappiness=0 debian:jessie true # memory cgroups is mounted
ba61c3bbe0198643994956aca074976e157ef2521425ac527b1a0b6016e61851
$ sudo umount /sys/fs/cgroup/memory  # umount memory cgroup for instance           
$ docker start ba61c3bbe0198643994956aca074976e157ef2521425ac527b1a0b6016e61851
Error response from daemon: Cannot start container ba61c3bbe0198643994956aca074976e157ef2521425ac527b1a0b6016e61851: [8] System error: no such directory for memory.swappiness.
Error: failed to start containers: [ba61c3bbe0198643994956aca074976e157ef2521425ac527b1a0b6016e61851]

```
This is happening because the hostConfig in the start API is only checked if it's the one provided in the request and not the actual container's hostConfig (see `verifyContainerSettings`)
Beaware also that `verifyContainerSettings` will set for instance `MemorySwappiness` to `nil` if it detects the system doesn't have memory cgroup enabled.
Should we check hostConfig before starting container (and show a nicer error and put MemorySwappiness in hostConfig to nil) or this libcontainer error is ok to show if you umount cgroups while running? I have a fix to check container's hostConfig on start also 

EDIT: Digging deeper, daemon.sysInfo is only populated on daemon start so even calling verifyContainerSettings in container start still shows the error above

ping @LK4D4 (I've seen libcontainer ignores cgroups path not found error in apply_systemd and try to write the file and erroring out above, wondering if that should be fixed as well)

### PR DESCRIPTION
**Pull request by runcom**
fix #15159 

tl;dr (more in commit messages)

sysinfo is initialized and stored in the daemon struct at daemon startup
runtime system settings changes aren't reflected in sysinfo daemon field
errors from libcontainer complaining about write to non existent cgroup path when it wasn't supposed to not even write there

now if a cgroup is umounted at runtime (memory for instance) and a container had memory-swappiness enabled and it's started again, the error from libcontainer is gone and in daemon logs you get

```
WARN[0108] Your kernel does not support cgroup memory limit: mountpoint for memory not found 
WARN[0108] Your kernel does not support memory swappiness capabilities, memory swappiness discarded. 
WARN[0108] IPv4 forwarding is disabled. Networking will not work 
WARN[0109] Your kernel does not support OOM notifications: There is no path for "memory" in state
```

**NOTE: warnings are not printed to stdout/stderr while starting a container (as it's happening for `run`) with wrong memory swappiness but only in daemon-logs.** This is because start return StatusNoContent so would not be good to break that behavior and print the warnings there and also this is more of a sysadmin thing who watches for wrong behaviors in daemon logs (please someone confirm this also)

I've made sysinfo checks at runtime and removed the field from the daemon struct
hostConfig is checked at container's start too

### CODE REVIEW

**Comment by icecrime** [comment]
Do you think we should consider it for 1.8.0 (cc @calavera)?

**Comment by runcom** [comment]
@icecrime yes I think, ppl don't usually umount cgroups at runtime or change sysctl settings but it may happen so better to have this check

I'll fix the panic/errors by tomorrow (too late here sorry)

**Comment by hqhq** [comment]
Looks good :)

**Comment by runcom** [comment]
@hqhq thanks, would be awesome if you give this patch a try :)

**Comment by hqhq** [comment]
@runcom After some tests, I do find some problems:
1. It would be better if we can return warnings for start api, maybe this can leave to next release;
2. If you use `-m` instead of `--memory-swappiness` in this scenario, you'll get:
```
Error response from daemon: You should always set the Memory limit when using Memoryswap limit, see usage.
```
because memorySwap was set to double of memory on create, we need to handle this.
3. This problem also affect other cgroup subsystems, such as cpu and cpuset, but I think we'll need #14015 

But I confirm that problem for `memory-swappiness` has been resolved :smile:

**Comment by runcom** [comment]
@hqhq perfect, thanks, I'll take care of your comments soon

**Comment by runcom** [comment]
@hqhq I've fixed various things, please confirm everything is working as expected if you have any chance

**Comment by calavera** [comment]
LGTM

File: daemon/daemon_unix.go [hunk 1]
```diff
@@ -131,20 +132,23 @@ func (daemon *Daemon) adaptContainerSettings(hostConfig *runconfig.HostConfig) {
...
-	var warnings []string
+	warnings := []string{}
+	sysInfo := sysinfo.New(false)
 
 	if hostConfig.LxcConf.Len() > 0 && !strings.Contains(daemon.ExecutionDriver().Name(), "lxc") {
 		return warnings, fmt.Errorf("Cannot use --lxc-conf with execdriver: %s", daemon.ExecutionDriver().Name())
 	}
+
+	// memory subsystem checks and adjustments
 	if hostConfig.Memory != 0 && hostConfig.Memory < 4194304 {
```
Comment (runcom) [inline daemon/daemon_unix.go]: This check can be moved after the cgroup subsystem check, no point in showing this if memory cgroup isn't supported

File: daemon/daemon_unix.go [hunk 2]
```diff
@@ -131,20 +132,23 @@ func (daemon *Daemon) adaptContainerSettings(hostConfig *runconfig.HostConfig) {
...
 		return warnings, fmt.Errorf("Minimum memory limit allowed is 4MB")
 	}
-	if hostConfig.Memory > 0 && !daemon.SystemConfig().MemoryLimit {
+	if hostConfig.Memory > 0 && !sysInfo.MemoryLimit {
 		warnings = append(warnings, "Your kernel does not support memory limit capabilities. Limitation discarded.")
 		logrus.Warnf("Your kernel does not support memory limit capabilities. Limitation discarded.")
 		hostConfig.Memory = 0
 	}
-	if hostConfig.Memory > 0 && hostConfig.MemorySwap != -1 && !daemon.SystemConfig().SwapLimit {
+	if hostConfig.MemorySwap != -1 && !sysInfo.SwapLimit {
```
Comment (hqhq) [inline daemon/daemon_unix.go]: It would be a problem, after this change, we'll get:
```
$ docker run -ti ubuntu bash
WARNING: Your kernel does not support swap limit capabilities, memory limited without swap.
root@e22ad4765a95:/#
```
That warning was not expected because we don't want to limit swap in this command.

File: daemon/daemon_unix.go [hunk 3]
```diff
@@ -131,20 +132,23 @@ func (daemon *Daemon) adaptContainerSettings(hostConfig *runconfig.HostConfig) {
...
+
+	// memory subsystem checks and adjustments
 	if hostConfig.Memory != 0 && hostConfig.Memory < 4194304 {
 		return warnings, fmt.Errorf("Minimum memory limit allowed is 4MB")
 	}
-	if hostConfig.Memory > 0 && !daemon.SystemConfig().MemoryLimit {
+	if hostConfig.Memory > 0 && !sysInfo.MemoryLimit {
 		warnings = append(warnings, "Your kernel does not support memory limit capabilities. Limitation discarded.")
 		logrus.Warnf("Your kernel does not support memory limit capabilities. Limitation discarded.")
 		hostConfig.Memory = 0
```
Comment (hqhq) [inline daemon/daemon_unix.go]: Maybe we can add `hostConfig.MemorySwap = -1` here.

File: daemon/daemon_unix.go (hunk 2 above)
Comment (runcom) [inline daemon/daemon_unix.go]: Ok there's some kind of problem with default value there, I'll fix

File: daemon/daemon_unix.go (hunk 3 above)
Comment (runcom) [inline daemon/daemon_unix.go]: You mean also setting memory swap to - 1 when memory limit is disabled in kernel?

Comment (hqhq) [inline daemon/daemon_unix.go]: That's one solution for the problem I mentioned.
```
$ docker run --name ttt -m 300M busybox true
$ sudo umount /sys/fs/cgroup/memory
$ docker start ttt
```
In this scenario, after docker run, we'll have hostConfig.Memory = 300M and hostConfig.MemorySwap = 600M, then we umount memory cgroup and try to start container, `sysinfo.MemoryLimit` would be false, so we set `hostConfig.Memory = 0`, but then we'll get `hostConfig.Memory == 0 && hostConfig.MemorySwap > 0` and end up error.

So here when we set `hostConfig.Memory = 0` we can also set `hostConfig.MemorySwap = -1`, it'll fix it, and because if we don't limit memory, we can't limit memory swap either, so the logic is right.

File: daemon/daemon_unix.go [hunk 4]
```diff
@@ -155,7 +160,7 @@ func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *runconfig.HostC
 	if hostConfig.Memory == 0 && hostConfig.MemorySwap > 0 {
 		return warnings, fmt.Errorf("You should always set the Memory limit when using Memoryswap limit, see usage.")
 	}
-	if hostConfig.MemorySwappiness != nil && !daemon.SystemConfig().MemorySwappiness {
+	if hostConfig.MemorySwappiness != nil && !sysInfo.MemorySwappiness {
```
Comment (runcom) [inline daemon/daemon_unix.go]: @hqhq now I get
```
./bundles/1.8.0-dev/binary/docker run -d debian:jessie true
WARNING: Your kernel does not support memory swappiness capabilities, memory swappiness discarded.
e8076e48660a6f7bded2ab67fb59a11c1878af79d78dff636957898e2803971e
```
but I suppose this is the correct behavior because Swappiness default value is -1

File: daemon/daemon_unix.go (hunk 3 above)
Comment (runcom) [inline daemon/daemon_unix.go]: updated

File: daemon/daemon_unix.go (hunk 4 above)
Comment (hqhq) [inline daemon/daemon_unix.go]: Emmm, then I think we should use:
`if hostConfig.MemorySwappiness != nil && *hostConfig.MemorySwappiness != -1 && !sysInfo.MemorySwappiness {`

-1 means don't set MemorySwappiness, let it inherit from parent, if we don't have Swappiness support, shouldn't show this warning.

Comment (runcom) [inline daemon/daemon_unix.go]: You're right, fixing

**Comment by calavera** [comment]
still LGTM

**Comment by runcom** [comment]
@calavera I also feel that the other PRs about cgroups checks should go into the release if this is going in (otherwise seems only some checks are going into the release and others related aren't). I'll add the milestone to those one but feel free to remove it if it doesn't sound good

**Comment by calavera** [comment]
let's evaluate them, I'm not completely sure which ones you're talking about

**Comment by calavera** [comment]
Closing this in favor of https://github.com/docker/docker/pull/15381 that's already rebased, so we can release RC 3. Thanks @runcom :metal:
